main.py

```diff
@@ -197,12 +198,14 @@ def do_match(cls, flow, args):
             return False
         for name in flow['flow']['match']:
             field_flow = flow['flow']['match'][name]
+            if name == 'dl_vlan':
+                field = args[name][-1] if name in args else 0
```

Comment:
@gretelliz, @Alopalao explanation is the behavior that's expected, if in the flow entry there's no `dl_vlan` set then it doesn't matter what `dl_vlan` is in the packet (args). There's a tiny diff when `dl_vlan` isn't set and when it's `dl_vlan: 0`, when it's not set.

Let's keep in match that we're performing a non-strict match of a packet args against a flow entry.

Reply:
> match anything (similar to any (dl_vlan="4096/4096") )
Is that correct?

It's not really "similar" to `any` -> "4096/4096", "4096/4096" will match a vlan within this inclusive [1, 4095] range.

Reply:
Thank you both. If something isn't clear, since there's lot of back and forth in the review I'll recommend that you both @gretelliz @Alopalao have a call to clarify, if you need my help, happy to also jump on it.

Reply:
You're highlighting the cases where it returns `True`, but considering when [`match_field_dl_vlan` returns `False`](https://github.com/kytos-ng/sdntrace_cp/blob/feature/type_trace/main.py#L203) is also pivotal to consider a mismatch, you'll probably want to focus on these ones too. The way that the is structured, whenever a matching func returns `False` it gives a semantics of a mismatch, let's keep that in mind for future reviews when new matching funcs are added to support other fields with masks too, they should follow this pattern that you're ensuring with only calling a matching func if [its match key field is defined in the flow](https://github.com/kytos-ng/sdntrace_cp/blob/feature/type_trace/main.py#L201). 

@gretelliz, regarding the examples you've given:

- `match_field_dl_vlan(A,A)` -> True. However, packet arg `a` shouldn't ever be `dl_vlan=0` in the first place, let's keep our core logic simple when applicable, `a` packet/arg either has a valid int VLAN from `1` to `4095` or `dl_vlan` isn't set (API validation will ensure this one day). If a client wants to send a untagged packet they have to send with `dl_vlan` missing (not set). `match_field_dl_vlan(B, A)` -> True. `match_field_dl_vlan(B, B)` should also be True, but notice that the func shouldn't even be called for this case (B, B) since flow match doesn't have `dl_vlan` key (see my prior paragraph, and let's follow the convention that you've already structured).
- `match_field_dl_vlan(B,_)` -> Wrong. If `_` means {A, B, C, D, E}, it would only return True if `b` is `A` or `B`. `match_field_dl_vlan(B,C)` would return False for instance.
- `match_field_dl_vlan(_,B)` -> We shouldn't have this call being performed, it goes back to the convention that I highlighted, if the flow entry doesn't have a `dl_vlan` why are we trying to match this field? It'll only make reasoning a bit more difficult to follow/understand, we could add early returns but it'd be redundant (also, we'll have additional matching funcs for other fields one day). Remember that code is read more times than it's written.
- `match_field_dl_vlan(E,C)` -> True. mask 4096 (flow mask) & 4095 (12-bit upper bound mask) will always be zero, so when the resulting logical and is performed with a value from [1-4095] and the flow value 4096 it'll be 0 == 0
- `match_field_dl_vlan(E,D)` -> Not always necessarily True, and in fact it's expected to result in False in some cases. Whenever there is a "num/num" (even for "4096/4096"), then you need to do the bitwise logical value & mask & 0xfff to confirm if it'll match. 

@Alopalao feel to confirm or complement if you wish.

Reply:
Ok, just note that `match_field_dl_vlan(A,A)` is no longer a case since the value of `vlan` in the packet (args) is an integer in [1, 4095].